File: experiments-local/experiment-16/runs/agent=omp_language=csharp_model=openrouter/anthropic/claude-opus-4.8_tooling=none/rep1/src/BrazilianSoccer.Core/Csv.cs
// =============================================================================
// Context: Brazilian Soccer MCP Server — minimal RFC 4180 CSV reader.
//
// The source CSVs are quoted inconsistently: some fields are double-quoted
// (Brasileirao), some are bare (novo_campeonato), team names contain embedded
// commas and parentheses ("Boavista Sport Club (antigo ...) - RJ"), and one file
// is UTF-8 BOM-prefixed (fifa_data). A dependency-free streaming parser keeps the
// loader honest about quoting and embedded delimiters without pulling in a CSV
// package. Returns rows as string arrays; the header row is exposed separately so
// loaders can index by column name.
// =============================================================================
using System.Text;

namespace BrazilianSoccer.Core;

public sealed class CsvTable
{
    private readonly Dictionary<string, int> _index;

    public IReadOnlyList<string> Header { get; }
    public IReadOnlyList<string[]> Rows { get; }

    public CsvTable(IReadOnlyList<string> header, IReadOnlyList<string[]> rows)
    {
        Header = header;
        Rows = rows;
        _index = new Dictionary<string, int>(StringComparer.OrdinalIgnoreCase);
        for (int i = 0; i < header.Count; i++)
            _index[header[i].Trim()] = i; // last-wins is fine; headers are unique here
    }

    /// <summary>Column ordinal for a header name, or -1 when absent.</summary>
    public int Col(string name) => _index.TryGetValue(name, out var i) ? i : -1;

    /// <summary>Cell value by column name, or null when the column/cell is missing or empty.</summary>
    public static string? Cell(string[] row, int col)
    {
        if (col < 0 || col >= row.Length)
            return null;
        var v = row[col];
        return string.IsNullOrWhiteSpace(v) ? null : v;
    }
}

public static class Csv
{
    /// <summary>Parse a CSV file from disk (UTF-8, BOM-tolerant) into a header + rows table.</summary>
    public static CsvTable Load(string path)
    {
        using var reader = new StreamReader(path, Encoding.UTF8, detectEncodingFromByteOrderMarks: true);
        var records = Parse(reader);
        if (records.Count == 0)
            return new CsvTable(Array.Empty<string>(), Array.Empty<string[]>());

        var header = records[0];
        // Strip a stray BOM that survived on the first header cell.
        if (header.Length > 0)
            header[0] = header[0].TrimStart('\uFEFF');

        var rows = new List<string[]>(records.Count - 1);
        for (int i = 1; i < records.Count; i++)
        {
            // Skip fully blank trailing lines.
            var r = records[i];
            if (r.Length == 1 && r[0].Length == 0)
                continue;
            rows.Add(r);
        }
        return new CsvTable(header, rows);
    }

    /// <summary>Streaming RFC 4180 parse honoring quotes, escaped quotes, and embedded newlines.</summary>
    private static List<string[]> Parse(TextReader reader)
    {
        var records = new List<string[]>();
        var fields = new List<string>();
        var field = new StringBuilder();
        bool inQuotes = false;
        bool fieldStarted = false;

        int ci;
        while ((ci = reader.Read()) >= 0)
        {
            char c = (char)ci;
            if (inQuotes)
            {
                if (c == '"')
                {
                    if (reader.Peek() == '"')
                    {
                        reader.Read();
                        field.Append('"');
                    }
                    else
                    {
                        inQuotes = false;
                    }
                }
                else
                {
                    field.Append(c);
                }
                continue;
            }

            switch (c)
            {
                case '"':
                    inQuotes = true;
                    fieldStarted = true;
                    break;
                case ',':
                    fields.Add(field.ToString());
                    field.Clear();
                    fieldStarted = false;
                    break;
                case '\r':
                    break; // handled with \n
                case '\n':
                    fields.Add(field.ToString());
                    field.Clear();
                    records.Add(fields.ToArray());
                    fields.Clear();
                    fieldStarted = false;
                    break;
                default:
                    field.Append(c);
                    fieldStarted = true;
                    break;
            }
        }

        // Flush trailing field/record if the file does not end with a newline.
        if (fieldStarted || field.Length > 0 || fields.Count > 0)
        {
            fields.Add(field.ToString());
            records.Add(fields.ToArray());
        }

        return records;
    }
}

File: experiments-local/experiment-17/runs/agent=omp_language=typescript_model=openrouter/anthropic/claude-opus-4.8_tooling=none/rep3/src/csv.ts
/**
 * Context
 * -------
 * Minimal, dependency-free RFC 4180 CSV parser for the Brazilian Soccer MCP
 * server. The provided Kaggle files mix quoting styles (some columns quoted,
 * some not), contain UTF-8 Portuguese text, a leading BOM (fifa_data.csv), and
 * embedded commas inside quoted fields. A streaming character-level parser is
 * used rather than a naive `split(",")` so those cases are handled correctly.
 *
 * Exports `parseCsv` -> array of row objects keyed by the header names.
 */

/** Parse CSV text into a list of objects keyed by the (trimmed) header row. */
export function parseCsv(text: string): Record<string, string>[] {
  const rows = parseRows(text);
  if (rows.length === 0) return [];

  const header = rows[0].map((h) => stripBom(h).trim());
  const out: Record<string, string>[] = [];

  for (let i = 1; i < rows.length; i++) {
    const cells = rows[i];
    // Skip fully empty trailing lines.
    if (cells.length === 1 && cells[0] === "") continue;
    const record: Record<string, string> = {};
    for (let c = 0; c < header.length; c++) {
      record[header[c]] = (cells[c] ?? "").trim();
    }
    out.push(record);
  }
  return out;
}

/** Parse raw CSV text into a 2D array of string cells. */
function parseRows(text: string): string[][] {
  const rows: string[][] = [];
  let field = "";
  let row: string[] = [];
  let inQuotes = false;
  let i = 0;
  const n = text.length;

  const pushField = () => {
    row.push(field);
    field = "";
  };
  const pushRow = () => {
    pushField();
    rows.push(row);
    row = [];
  };

  while (i < n) {
    const ch = text[i];

    if (inQuotes) {
      if (ch === '"') {
        if (text[i + 1] === '"') {
          field += '"';
          i += 2;
          continue;
        }
        inQuotes = false;
        i++;
        continue;
      }
      field += ch;
      i++;
      continue;
    }

    if (ch === '"') {
      inQuotes = true;
      i++;
      continue;
    }
    if (ch === ",") {
      pushField();
      i++;
      continue;
    }
    if (ch === "\r") {
      // Handle CRLF and lone CR.
      if (text[i + 1] === "\n") i++;
      pushRow();
      i++;
      continue;
    }
    if (ch === "\n") {
      pushRow();
      i++;
      continue;
    }
    field += ch;
    i++;
  }

  // Flush trailing field/row if the file did not end with a newline.
  if (field.length > 0 || row.length > 0) pushRow();

  return rows;
}

function stripBom(s: string): string {
  return s.charCodeAt(0) === 0xfeff ? s.slice(1) : s;
}
